File: 7.binary-search-tree2/동규/4.kth-smallest-element-in-a-sorted-matrix.js
// https://leetcode.com/problems/kth-smallest-element-in-a-sorted-matrix/

// sol 1
var kthSmallest = function (matrix, k) {
  const arr = matrix.reduce((acc, cur) => {
    return acc.concat(cur);
  });
  arr.sort((a, b) => a - b);
  return arr[k - 1];
};

// var kthSmallest = function (matrix, k) {
//   let left = matrix[0][0];
//   let right = matrix[matrix.length - 1][matrix.length - 1];

//   while (left <= right) {
//     const middle = (left + right) / 2 + left;
//     const potentialMatch = matrix[middle];
//     if (k === potentialMatch) {
//       return middle;
//     } else if (k < potentialMatch) {
//       right = middle - 1;
//     } else {
//       left = middle + 1;
//     }
//   }
//   return left;
// };

File: 7.binary-search-tree2/지연/4.kth-smallest-element-in-a-sorted-matrix.py
# https://leetcode.com/problems/kth-smallest-element-in-a-sorted-matrix/

from typing import List


class Solution:
    def kthSmallest(self, matrix: List[List[int]], k: int) -> int:
        row_len, col_len = len(matrix), len(matrix[0])
        new_list = []

        if(len(matrix) <= 1):
            return matrix[0][0]
        for row in range(row_len):
            for col in range(col_len):
                new_list.append(matrix[row][col])

        new_list.sort()
        return new_list[k - 1]
